hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java

```diff
@@ -53,18 +53,18 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
 
     @Override
     public void beforeRun() {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
-        ringbuffer.checkBlockableReadSequence(startSequence);
     }
 
     @Override
     public boolean shouldWait() {
         if (resultSet == null) {
-            resultSet = new ReadResultSetImpl<O, O>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
+            resultSet = new ReadResultSetImpl<>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
             sequence = startSequence;
         }
 
         RingbufferContainer ringbuffer = getRingBufferContainer();
+        sequence = ringbuffer.clampReadSequenceToBounds(sequence);
+
         if (minSize == 0) {
             if (!ringbuffer.shouldWait(sequence)) {
                 readMany(ringbuffer);
```

Comment:
Isn't this basically the same thing as your previous comment? Am I missing something?

Reply:
No, it is different line. I mean that `readMany` can throw `StaleSequenceException`.

Reply:
Ahh, ok. I understand now. We probably can't/shouldn't make `RingbufferContainer.readMany` not throw `StaleSequenceException`, because it's used by the `EventJournal` too, but in `ReadManyOperation.shouldWait`, as far as I can tell, it can actually throw the exception only if the store is not enabled. I would say we should just specify that in the docs. Will do that, let me know if you have a different opinion.

Reply:
Actually no, I'm mistaken, it won't throw the exception even then (when the store is enabled). Yes clamping is not happening in that case, but the sequence won't be considered stale, the exception won't be thrown, it will instead attempt to load it from the store (and just return `null` if it can't).

Reply:
Even if you clamped the sequence number, until you do the `RingbufferContainer.readMany` call, is it not possible that another thread did a number of readMany's and your clamped sequence number became stale again?

Reply:
It is correct that we want to avoid the `StaleSequenceException` in this PR but we want to do this by clamping and relying that we pass any subsequent checks, instead of removing those checks.

With that in mind, I don't see a problem here, unless there is some concurrent mutation on the ringbuffer, same as in the previous case - https://github.com/hazelcast/hazelcast/pull/16303#discussion_r363743868 and I agree with @jbartok . 

AFAIK the entire `ReadManyOperation#shouldWait` is executed on the partition thread in it's entirety and items can be added to the same ringbuffer only once the partition thread is freed.

So, if clamping will restrict the sequence to [headSequence, tailSequence] (or less than headSequence if store is enabled), since there cannot be any concurrent mutation on the ringbuffer, since no other thread than the partition thread can invoke `readMany` and since `checkBlockableReadSequence`  throws an exception only if `sequence > tailSequence() + 1` (which is not possible because of clamping) or if `sequence < headSequence() && !store.isEnabled()` (which again is not possible because of clamping), I don't see how an exception can be thrown here.

It is possible, though, that the `shouldWait()` method may return (because there are not enough items) after which new items might be added, and the `shouldWait()` is executed again. This, again, is not an issue since we clamp every time we execute the method.

@ihsandemir can you elaborate with a specific scenario and thread scheduling how we might encounter an exception here?

Reply:
Ok, I understand now. I was assuming that concurrent mutation was possible but since it runs on the certain partition thread, I think that will not be a problem.